# Projekt DS & KI
Gruppenmitglieder: Clemens Fleischhauer, Nils Groetelaers, Nikita Helwer, Carl Placek, Julian Krupp

# Unser Datensatz:

https://archive.ics.uci.edu/dataset/186/wine+quality

# SMART-Ziel:

Wir wollen untersuchen, welchen Einfluss der Alkohol-Prozentgehalt auf die Weinqualität hat. Der UCI-Weindatensatz mit ca. 5000 Weißwein- und ca. 1500 Rotweinproben liefert ausreichend Daten, um erste Schlüsse zu ziehen. Unser Ziel ist die Analyse von Korrelationen und weiteren Zusammenhängen, sowie deren Schilderung und Dokumentation innerhalb der vorgegebenen Zeit (bis 23.11.).

# Quellen:

Titel: Modellierung der Weinqualität anhand physiochemischer Eigenschaften

https://cs229.stanford.edu/proj2019aut/data/assignment_308832_raw/25895690.pdf

Erkenntnisse aus der Quelle: Anhand der vier Größen chlorides, total sulfur dioxide, fixed acidity und pH oder chlorides, total sulfur dioxide, volatile acidity und free sulfur dioxide lässt sich die Weinqualität mithilfe eines Neural Network Models mit einer 99%-igen Sicherheit voraussagen. 

Wir sehen also: Alkohol scheint sich im Vergleich zu anderen Parametern nicht so gut zu eignen, um die Weinqualität vorauszusagen. Daher hat Alkohol vermutlich einen weniger starken Einfluss auf die Weinqualität als die anderen zur Vorhersage benutzten Parameter.

# Aufgabe 1:

In [1]:
# Datensatz Weißwein
import pandas as pd
ww = pd.read_csv('winequality-white.csv', sep=';')

print(ww.head())

print("=====================================================================================================================================>")

print(ww.columns) # 12 Spalten

print("=====================================================================================================================================>")

print(ww.index)   # 4898 Zeilen

print("=====================================================================================================================================>")

ww.describe()

   fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.0              0.27         0.36            20.7      0.045   
1            6.3              0.30         0.34             1.6      0.049   
2            8.1              0.28         0.40             6.9      0.050   
3            7.2              0.23         0.32             8.5      0.058   
4            7.2              0.23         0.32             8.5      0.058   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 45.0                 170.0   1.0010  3.00       0.45   
1                 14.0                 132.0   0.9940  3.30       0.49   
2                 30.0                  97.0   0.9951  3.26       0.44   
3                 47.0                 186.0   0.9956  3.19       0.40   
4                 47.0                 186.0   0.9956  3.19       0.40   

   alcohol  quality  
0      8.8        6  
1      9.5        6  
2     10.1        6 

fixed acidity  volatile acidity  citric acid  residual sugar  \
count    4898.000000       4898.000000  4898.000000     4898.000000   
mean        6.854788          0.278241     0.334192        6.391415   
std         0.843868          0.100795     0.121020        5.072058   
min         3.800000          0.080000     0.000000        0.600000   
25%         6.300000          0.210000     0.270000        1.700000   
50%         6.800000          0.260000     0.320000        5.200000   
75%         7.300000          0.320000     0.390000        9.900000   
max        14.200000          1.100000     1.660000       65.800000   

         chlorides  free sulfur dioxide  total sulfur dioxide      density  \
count  4898.000000          4898.000000           4898.000000  4898.000000   
mean      0.045772            35.308085            138.360657     0.994027   
std       0.021848            17.007137             42.498065     0.002991   
min       0.009000             2.000000              9.000000     0.987110   
25%       0.036000            23.000000            108.000000     0.991723   
50%       0.043000            34.000000            134.000000     0.993740   
75%       0.050000            46.000000            167.000000     0.996100   
max       0.346000           289.000000            440.000000     1.038980   

                pH    sulphates      alcohol      quality  
count  4898.000000  4898.000000  4898.000000  4898.000000  
mean      3.188267     0.489847    10.514267     5.877909  
std       0.151001     0.114126     1.230621     0.885639  
min       2.720000     0.220000     8.000000     3.000000  
25%       3.090000     0.410000     9.500000     5.000000  
50%       3.180000     0.470000    10.400000     6.000000  
75%       3.280000     0.550000    11.400000     6.000000  
max       3.820000     1.080000    14.200000     9.000000

In [2]:
# Datensatz Rotwein
import pandas as pd
rw = pd.read_csv('winequality-red.csv', sep=';')

print(rw.head())

print("=====================================================================================================================================>")

print(rw.columns) # 12 Spalten

print("=====================================================================================================================================>")

print(rw.index)   # 1599 Zeilen

print("=====================================================================================================================================>")

rw.describe()

   fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  quality  
0      9.4        5  
1      9.8        5  
2      9.8        5 

fixed acidity  volatile acidity  citric acid  residual sugar  \
count    1599.000000       1599.000000  1599.000000     1599.000000   
mean        8.319637          0.527821     0.270976        2.538806   
std         1.741096          0.179060     0.194801        1.409928   
min         4.600000          0.120000     0.000000        0.900000   
25%         7.100000          0.390000     0.090000        1.900000   
50%         7.900000          0.520000     0.260000        2.200000   
75%         9.200000          0.640000     0.420000        2.600000   
max        15.900000          1.580000     1.000000       15.500000   

         chlorides  free sulfur dioxide  total sulfur dioxide      density  \
count  1599.000000          1599.000000           1599.000000  1599.000000   
mean      0.087467            15.874922             46.467792     0.996747   
std       0.047065            10.460157             32.895324     0.001887   
min       0.012000             1.000000              6.000000     0.990070   
25%       0.070000             7.000000             22.000000     0.995600   
50%       0.079000            14.000000             38.000000     0.996750   
75%       0.090000            21.000000             62.000000     0.997835   
max       0.611000            72.000000            289.000000     1.003690   

                pH    sulphates      alcohol      quality  
count  1599.000000  1599.000000  1599.000000  1599.000000  
mean      3.311113     0.658149    10.422983     5.636023  
std       0.154386     0.169507     1.065668     0.807569  
min       2.740000     0.330000     8.400000     3.000000  
25%       3.210000     0.550000     9.500000     5.000000  
50%       3.310000     0.620000    10.200000     6.000000  
75%       3.400000     0.730000    11.100000     6.000000  
max       4.010000     2.000000    14.900000     8.000000

## Anzahl Zeilen und Spalten und deren Bedeutung

### Datensatz Rotwein

Anzahl Zeilen: 1599

Anzahl Spalten: 12

### Datensatz Weißwein

Anzahl Zeilen: 4898 

Anzahl Spalten: 12

### Spalten (gleich für beide Datensätze):

"fixed acidity" = g/L, stabile Säuren, bestimmen den pH-Wert und Geschmack

"volatile acidity" = g/L, in geringen Mengen für Aroma gewünscht, zu viel ist ein Weinfehler

"citric acid" =  g/L, Teil der fixen Säuren, trägt zur Frische bei

"residual sugar" = g/L, unvergorener Zucker im Wein

"chlorides" = g/L, Salze, die den Salzgehalt beeinflussen

"free sulfur dioxide" = mg/L, schützt Wein vor Oxidation und unerwünschten Mikroorganismen

"total sulfur dioxide" = mg/L, Summe aus freiem und gebundenem SO2, Gesamtschutz des Weins

"density" = g/cm3, Dichte des Weins

"pH" = von 1 bis 14, 1 - 7 heißt sauer, 7 - 14 ist basisch

"sulphates" = g/L, Salze der Schwefelsäure, können die Stabilität und Konservierung des Weins beeinflussen

"alcohol" = Alkoholgehalt in Prozent

"quality" 

### Was fällt auf? ww.describe()

Ausreißer könnten fixed acidity, residual sugar, free sulfur dioxide und total sulfur dioxide sein; der maximale Wert ist teils ein erhebliches Vielfaches des 75%-Perzentil-Wert

keine fehlenden Werte (bereinigt)

# Aufgabe 2:

## Welche Fragestellungen für Ihren Datensatz können Sie mit der Indizierung und Manipulation                                                                                 bzw. Gruppierungsfunktion lösen?

Wie viel Prozent der Weine haben eine Qualität von 8 oder höher? (8 oder höher entspricht einer sehr hohen Qualität)

-> Nur 180 Weißweine und 18 Rotweine haben eine Qualität von 8 oder höher. Das entspricht 198/6497 Weinproben (~ 3,05% aller getesteten Weine).

In [3]:
# Datensatz Weißwein
ww[ww.quality>=8] # 180 Weißweine haben eine Qualität von 8 oder höher. 

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
17              6.2             0.660         0.48             1.2      0.029   
20              6.2             0.660         0.48             1.2      0.029   
22              6.8             0.260         0.42             1.7      0.049   
68              6.7             0.230         0.31             2.1      0.046   
74              6.7             0.230         0.31             2.1      0.046   
...             ...               ...          ...             ...        ...   
4765            5.2             0.300         0.34             1.5      0.038   
4766            6.4             0.320         0.25             5.0      0.055   
4786            4.4             0.320         0.39             4.3      0.030   
4787            3.9             0.225         0.40             4.2      0.030   
4802            5.8             0.280         0.34             2.2      0.037   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
17                   29.0                  75.0  0.98920  3.33       0.39   
20                   29.0                  75.0  0.98920  3.33       0.39   
22                   41.0                 122.0  0.99300  3.47       0.48   
68                   30.0                  96.0  0.99260  3.33       0.64   
74                   30.0                  96.0  0.99260  3.33       0.64   
...                   ...                   ...      ...   ...        ...   
4765                 18.0                  96.0  0.98942  3.56       0.48   
4766                 28.0                 138.0  0.99171  3.27       0.50   
4786                 31.0                 127.0  0.98904  3.46       0.36   
4787                 29.0                 118.0  0.98900  3.57       0.36   
4802                 24.0                 125.0  0.98986  3.36       0.33   

      alcohol  quality  
17       12.8        8  
20       12.8        8  
22       10.5        8  
68       10.7        8  
74       10.7        8  
...       ...      ...  
4765     13.0        8  
4766     12.4        8  
4786     12.8        8  
4787     12.8        8  
4802     12.8        8  

[180 rows x 12 columns]

In [4]:
# Datensatz Rotwein
rw[rw.quality>=8] # Nur 18 Rotweine haben eine Qualität von 8 oder höher.

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
267             7.9              0.35         0.46             3.6      0.078   
278            10.3              0.32         0.45             6.4      0.073   
390             5.6              0.85         0.05             1.4      0.045   
440            12.6              0.31         0.72             2.2      0.072   
455            11.3              0.62         0.67             5.2      0.086   
481             9.4              0.30         0.56             2.8      0.080   
495            10.7              0.35         0.53             2.6      0.070   
498            10.7              0.35         0.53             2.6      0.070   
588             5.0              0.42         0.24             2.0      0.060   
828             7.8              0.57         0.09             2.3      0.065   
1061            9.1              0.40         0.50             1.8      0.071   
1090           10.0              0.26         0.54             1.9      0.083   
1120            7.9              0.54         0.34             2.5      0.076   
1202            8.6              0.42         0.39             1.8      0.068   
1269            5.5              0.49         0.03             1.8      0.044   
1403            7.2              0.33         0.33             1.7      0.061   
1449            7.2              0.38         0.31             2.0      0.056   
1549            7.4              0.36         0.30             1.8      0.074   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
267                  15.0                  37.0  0.99730  3.35       0.86   
278                   5.0                  13.0  0.99760  3.23       0.82   
390                  12.0                  88.0  0.99240  3.56       0.82   
440                   6.0                  29.0  0.99870  2.88       0.82   
455                   6.0                  19.0  0.99880  3.22       0.69   
481                   6.0                  17.0  0.99640  3.15       0.92   
495                   5.0                  16.0  0.99720  3.15       0.65   
498                   5.0                  16.0  0.99720  3.15       0.65   
588                  19.0                  50.0  0.99170  3.72       0.74   
828                  34.0                  45.0  0.99417  3.46       0.74   
1061                  7.0                  16.0  0.99462  3.21       0.69   
1090                 42.0                  74.0  0.99451  2.98       0.63   
1120                  8.0                  17.0  0.99235  3.20       0.72   
1202                  6.0                  12.0  0.99516  3.35       0.69   
1269                 28.0                  87.0  0.99080  3.50       0.82   
1403                  3.0                  13.0  0.99600  3.23       1.10   
1449                 15.0                  29.0  0.99472  3.23       0.76   
1549                 17.0                  24.0  0.99419  3.24       0.70   

      alcohol  quality  
267      12.8        8  
278      12.6        8  
390      12.9        8  
440       9.8        8  
455      13.4        8  
481      11.7        8  
495      11.0        8  
498      11.0        8  
588      14.0        8  
828      12.7        8  
1061     12.5        8  
1090     11.8        8  
1120     13.1        8  
1202     11.7        8  
1269     14.0        8  
1403     10.0        8  
1449     11.3        8  
1549     11.4        8

Was ist der durchschnittliche Alkoholgehalt für jede Weinqualität?

- genauer in Ausgabe einsehbar (unten)
- tatsächlich haben qualitativ hochwertige Weine auch höheren Alkoholanteil
- geringere Mengen an hochqualitativen Weinen, d. h. Ausreißer haben stärkere Wirkung

In [5]:
# Datensatz Weißwein
ww_mean_alcohol_for_each_quality = ww.groupby('quality')['alcohol'].mean()

print(f"Durchschnittlicher Alkoholgehalt für jede Qualitätsstufe (Stufe 1, 2, 10 sind im Datensatz nicht vergeben):\n"
      f"\n{ww_mean_alcohol_for_each_quality}")

Durchschnittlicher Alkoholgehalt für jede Qualitätsstufe (Stufe 1, 2, 10 sind im Datensatz nicht vergeben):

quality
3    10.345000
4    10.152454
5     9.808840
6    10.575372
7    11.367936
8    11.636000
9    12.180000
Name: alcohol, dtype: float64


In [6]:
# Datensatz Rotwein
rw_mean_alcohol_for_each_quality = rw.groupby('quality')['alcohol'].mean()

print(f"Durchschnittlicher Alkoholgehalt für jede Qualitätsstufe (Stufe 1, 2, 9, 10 sind im Datensatz nicht vergeben):\n"
      f"\n{rw_mean_alcohol_for_each_quality}")

Durchschnittlicher Alkoholgehalt für jede Qualitätsstufe (Stufe 1, 2, 9, 10 sind im Datensatz nicht vergeben):

quality
3     9.955000
4    10.265094
5     9.899706
6    10.629519
7    11.465913
8    12.094444
Name: alcohol, dtype: float64


## Formulieren Sie Hypothesen.

Der Wert für citric acid muss stets geringer sein als der Wert für fixed acidity.

-> trifft zu, Hypothese bestätigt (siehe unten)

-> fixed acidity - citric acid ist stets größer 0

In [7]:
# Datensatz Weißwein
ww_fa_minus_ca = ww['fixed acidity']-ww['citric acid']
# positiv für alle Proben (logisch, da Zitronensäure eine fixe Säure ist)

print(f"Fixe Acidität minus Zitronensäure (Weißwein):\n"
      f"\n{ww_fa_minus_ca}")

Fixe Acidität minus Zitronensäure (Weißwein):

0       6.64
1       5.96
2       7.70
3       6.88
4       6.88
        ... 
4893    5.91
4894    6.24
4895    6.31
4896    5.20
4897    5.62
Length: 4898, dtype: float64


In [8]:
# Datensatz Rotwein
rw_fa_minus_ca = rw['fixed acidity']-rw['citric acid']
# positiv für alle Proben (logisch, da Zitronensäure eine fixe Säure ist)

print(f"Fixe Acidität minus Zitronensäure (Rotwein):\n"
      f"\n{rw_fa_minus_ca}")

Fixe Acidität minus Zitronensäure (Rotwein):

0        7.40
1        7.80
2        7.76
3       10.64
4        7.40
        ...  
1594     6.12
1595     5.80
1596     6.17
1597     5.78
1598     5.53
Length: 1599, dtype: float64


Hohe volatile acidity weißt auf verdorbenen Wein hin, folglich ist die Qualität niedriger.

- trifft nicht eindeutig zu, Differenz zum Durchschnitt eher gering (siehe unten)
- durchschnittliche Qualität bei hoher volatile acidity (über oder gleich dem 75% Perzentil) liegt in beiden Fällen (bei Rot- und Weißwein) ein wenig unter dem Durchschnitt:
- Weißwein: durchschnittl. Qualität = 5.88, durchschnittl. Qualität bei hoher volatile acidity = 5.72
- Rotwein: durchschnittl. Qualität = 5.64, durchschnittl. Qualität bei hoher volatile acidity = 5.27

Niedrige volatile acidity weißt auf fehlendes Aroma hin, folglich ist die Qualität niedriger.

- trifft auch nicht eindeutig zu, Differenz zum Durchschnitt eher gering (siehe unten)
- durchschnittliche Qualität bei niedriger volatile acidity (unter oder gleich dem 25% Perzentil) liegt in beiden Fällen (bei Rot- und Weißwein) ein wenig über dem Durchschnitt:
- Weißwein: durchschnittl. Qualität = 5.88, durchschnittl. Qualität bei niedriger volatile acidity = 6.11
- Rotwein: durchschnittl. Qualität = 5.64, durchschnittl. Qualität bei niedriger volatile acidity = 6.08

In [9]:
# Datensatz Weißwein
mean_weisswein_high_va = ww[ww['volatile acidity']>=0.32]['quality'].mean() 
# gibt Durchschnittsqualität für alle Weißweinproben mit volatile acidity >= 0.32 g/L aus

mean_weisswein_low_va = ww[ww['volatile acidity']<=0.21]['quality'].mean() 
# gibt Durchschnittsqualität für alle Weißweinproben mit volatile acidity <= 0.21 g/L aus

print(f"Durchschnittliche Qualität bei hoher volatiler Acidität (Weißwein): {mean_weisswein_high_va}\n")

print(f"Durchschnittliche Qualität bei niedriger volatiler Acidität (Weißwein): {mean_weisswein_low_va}")

Durchschnittliche Qualität bei hoher volatiler Acidität (Weißwein): 5.723388848660391

Durchschnittliche Qualität bei niedriger volatiler Acidität (Weißwein): 6.109589041095891


In [10]:
# Datensatz Rotwein
mean_rotwein_high_va = rw[rw['volatile acidity']>=0.64]['quality'].mean() 
# gibt Durchschnittsqualität für alle Rotweinproben mit volatile acidity >= 0.64 g/L aus

mean_rotwein_low_va = rw[rw['volatile acidity']<=0.39]['quality'].mean() 
# gibt Durchschnittsqualität für alle Rotweinproben mit volatile acidity <= 0.39 g/L aus

print(f"Durchschnittliche Qualität bei hoher volatiler Acidität (Rotwein): {mean_rotwein_high_va}\n")

print(f"Durchschnittliche Qualität bei niedriger volatiler Acidität (Rotwein): {mean_rotwein_low_va}")

Durchschnittliche Qualität bei hoher volatiler Acidität (Rotwein): 5.2754342431761785

Durchschnittliche Qualität bei niedriger volatiler Acidität (Rotwein): 6.078817733990148


# Aufgabe 3:

## Bestimmen Sie Lage- und Streuungsmaße der Merkmale in Ihrem Datensatz. Verwenden Sie df.describe()

In [11]:
# Datensatz Weißwein
ww.describe()

fixed acidity  volatile acidity  citric acid  residual sugar  \
count    4898.000000       4898.000000  4898.000000     4898.000000   
mean        6.854788          0.278241     0.334192        6.391415   
std         0.843868          0.100795     0.121020        5.072058   
min         3.800000          0.080000     0.000000        0.600000   
25%         6.300000          0.210000     0.270000        1.700000   
50%         6.800000          0.260000     0.320000        5.200000   
75%         7.300000          0.320000     0.390000        9.900000   
max        14.200000          1.100000     1.660000       65.800000   

         chlorides  free sulfur dioxide  total sulfur dioxide      density  \
count  4898.000000          4898.000000           4898.000000  4898.000000   
mean      0.045772            35.308085            138.360657     0.994027   
std       0.021848            17.007137             42.498065     0.002991   
min       0.009000             2.000000              9.000000     0.987110   
25%       0.036000            23.000000            108.000000     0.991723   
50%       0.043000            34.000000            134.000000     0.993740   
75%       0.050000            46.000000            167.000000     0.996100   
max       0.346000           289.000000            440.000000     1.038980   

                pH    sulphates      alcohol      quality  
count  4898.000000  4898.000000  4898.000000  4898.000000  
mean      3.188267     0.489847    10.514267     5.877909  
std       0.151001     0.114126     1.230621     0.885639  
min       2.720000     0.220000     8.000000     3.000000  
25%       3.090000     0.410000     9.500000     5.000000  
50%       3.180000     0.470000    10.400000     6.000000  
75%       3.280000     0.550000    11.400000     6.000000  
max       3.820000     1.080000    14.200000     9.000000

In [12]:
# Datensatz Rotwein
rw.describe()

fixed acidity  volatile acidity  citric acid  residual sugar  \
count    1599.000000       1599.000000  1599.000000     1599.000000   
mean        8.319637          0.527821     0.270976        2.538806   
std         1.741096          0.179060     0.194801        1.409928   
min         4.600000          0.120000     0.000000        0.900000   
25%         7.100000          0.390000     0.090000        1.900000   
50%         7.900000          0.520000     0.260000        2.200000   
75%         9.200000          0.640000     0.420000        2.600000   
max        15.900000          1.580000     1.000000       15.500000   

         chlorides  free sulfur dioxide  total sulfur dioxide      density  \
count  1599.000000          1599.000000           1599.000000  1599.000000   
mean      0.087467            15.874922             46.467792     0.996747   
std       0.047065            10.460157             32.895324     0.001887   
min       0.012000             1.000000              6.000000     0.990070   
25%       0.070000             7.000000             22.000000     0.995600   
50%       0.079000            14.000000             38.000000     0.996750   
75%       0.090000            21.000000             62.000000     0.997835   
max       0.611000            72.000000            289.000000     1.003690   

                pH    sulphates      alcohol      quality  
count  1599.000000  1599.000000  1599.000000  1599.000000  
mean      3.311113     0.658149    10.422983     5.636023  
std       0.154386     0.169507     1.065668     0.807569  
min       2.740000     0.330000     8.400000     3.000000  
25%       3.210000     0.550000     9.500000     5.000000  
50%       3.310000     0.620000    10.200000     6.000000  
75%       3.400000     0.730000    11.100000     6.000000  
max       4.010000     2.000000    14.900000     8.000000

(50% = Median, std [standard deviation] = Standartabweichung, mean = Arithmetisches Mittel)

## Haben Sie diese Ergebnisse erwartet?

Im Allgemeinen sind diese Ergebnisse zu erwarten.

## Spannende Erkenntnisse:

- pH ist bei rund 3, was für ein Getränk relativ sauer ist. Ein saurer pH-Wert (kleiner als 7) bei Wein ist allerdings nicht verwunderlich.
- Geringer Effekt des Alkoholgehalts auf die Dichte: Arithmetisches Mittel der Dichte ist rund 1, obwohl durchschnittlich etwa 10% Alkohol im Wein enthalten sind. Die Dichte von Alkohol ist aber bei Raumtemperatur um 20% geringer als die Dichte von Wasser (Ethanol: 0,8 g/cm^3, Wasser: 1 g/cm^3).

## Liegen Mittelwert und Median weit auseinander? Wie groß ist die Streuung?

Median und Mittelwert stehen in fast allen Fällen (bei Weiß- und Rotwein) nah beieinander, was auf eine geringe Streuung hinweist. Ausnahmen sind residual sugar und beide Kategorien des sulfur dioxide: Bei Plan zum späteren Verkauf/Konsum muss sulfur dioxide höher sein, um den Wein lagern zu können. Zucker schwankt aufgrund der Weinsorten stärker, denn ein lieblicher oder halbtrockener Wein enthält mehr Zucker als ein trockener Wein. Leider ist die genauere Weinkategorie nicht im Datensatz angegeben.

# Aufgabe 4:

## Skalenniveaus der Merkmale (Bedeutung von oben ergänzen)

| Merkmal | Skalenniveau | Bedeutung | Einheit |
|---------------|---------------|---------------|---------------|
| fixed acidity | kardinal | stabile Säuren, bestimmen den pH-Wert und Geschmack | g/L |
| volatile acidity | kardinal | in geringen Mengen für Aroma gewünscht, zu viel ist ein Weinfehler | g/L |
| citric acid | kardinal | Teil der fixen Säuren, trägt zur Frische bei | g/L |
| residual sugar | kardinal | unvergorener Zucker im Wein | g/L |
| chlorides | kardinal | Salze, die den Salzgehalt beeinflussen | g/L |
| free sulfur dioxide | kardinal | schützt Wein vor Oxidation und unerwünschten Mikroorganismen | mg/L |
| total sulfur dioxide | kardinal | Summe aus freiem und gebundenem SO2, Gesamtschutz des Weins | mg/L |
| density | kardinal | Dichte des Weins | g/cm^3 |
| pH | kardinal | 1 - 7 heißt sauer, 7 - 14 ist basisch | 1 - 14 |
| sulphates | kardinal | Salze der Schwefelsäure, können die Stabilität und Konservierung des Weins beeinflussen | g/L |
| alcohol | kardinal | Alkoholgehalt | % |
| quality | ordinal / kardinal, wir betrachten es als kardinal | Qualität / Geschmack | 1 - 10 |

# Aufgabe 5:

## Welche Korellationskoeffizienten möchten sie anwenden?
- Bravais-Pearson-Korrelationskoeffizient, da wir hauptsächlich bzw. nur kardinal skalierte Variablen untersuchen

## Prüfen Sie mögliche Zusammenhänge. Waren die Ergebnisse zu erwarten?

In [13]:
kardinal_columns = ['fixed acidity', 'volatile acidity', 'citric acid', 'residual sugar', 'chlorides', 'free sulfur dioxide', 
                   'total sulfur dioxide', 'density', 'pH', 'sulphates', 'alcohol', 'quality']

In [14]:
# Datensatz Weißwein
ww[kardinal_columns].corr()

fixed acidity  volatile acidity  citric acid  \
fixed acidity              1.000000         -0.022697     0.289181   
volatile acidity          -0.022697          1.000000    -0.149472   
citric acid                0.289181         -0.149472     1.000000   
residual sugar             0.089021          0.064286     0.094212   
chlorides                  0.023086          0.070512     0.114364   
free sulfur dioxide       -0.049396         -0.097012     0.094077   
total sulfur dioxide       0.091070          0.089261     0.121131   
density                    0.265331          0.027114     0.149503   
pH                        -0.425858         -0.031915    -0.163748   
sulphates                 -0.017143         -0.035728     0.062331   
alcohol                   -0.120881          0.067718    -0.075729   
quality                   -0.113663         -0.194723    -0.009209   

                      residual sugar  chlorides  free sulfur dioxide  \
fixed acidity               0.089021   0.023086            -0.049396   
volatile acidity            0.064286   0.070512            -0.097012   
citric acid                 0.094212   0.114364             0.094077   
residual sugar              1.000000   0.088685             0.299098   
chlorides                   0.088685   1.000000             0.101392   
free sulfur dioxide         0.299098   0.101392             1.000000   
total sulfur dioxide        0.401439   0.198910             0.615501   
density                     0.838966   0.257211             0.294210   
pH                         -0.194133  -0.090439            -0.000618   
sulphates                  -0.026664   0.016763             0.059217   
alcohol                    -0.450631  -0.360189            -0.250104   
quality                    -0.097577  -0.209934             0.008158   

                      total sulfur dioxide   density        pH  sulphates  \
fixed acidity                     0.091070  0.265331 -0.425858  -0.017143   
volatile acidity                  0.089261  0.027114 -0.031915  -0.035728   
citric acid                       0.121131  0.149503 -0.163748   0.062331   
residual sugar                    0.401439  0.838966 -0.194133  -0.026664   
chlorides                         0.198910  0.257211 -0.090439   0.016763   
free sulfur dioxide               0.615501  0.294210 -0.000618   0.059217   
total sulfur dioxide              1.000000  0.529881  0.002321   0.134562   
density                           0.529881  1.000000 -0.093591   0.074493   
pH                                0.002321 -0.093591  1.000000   0.155951   
sulphates                         0.134562  0.074493  0.155951   1.000000   
alcohol                          -0.448892 -0.780138  0.121432  -0.017433   
quality                          -0.174737 -0.307123  0.099427   0.053678   

                       alcohol   quality  
fixed acidity        -0.120881 -0.113663  
volatile acidity      0.067718 -0.194723  
citric acid          -0.075729 -0.009209  
residual sugar       -0.450631 -0.097577  
chlorides            -0.360189 -0.209934  
free sulfur dioxide  -0.250104  0.008158  
total sulfur dioxide -0.448892 -0.174737  
density              -0.780138 -0.307123  
pH                    0.121432  0.099427  
sulphates            -0.017433  0.053678  
alcohol               1.000000  0.435575  
quality               0.435575  1.000000

### erwähnenswerte Zusammenhänge im Weißwein - Datensatz:
- fixed acidity - citric acid (0.29, denn citric acid ist ein Teil der fixed acidity)
- fixed acidity - pH (-0.43, denn Menge an Säure im Wein bestimmt pH-Wert)
- residual sugar und free sulfur dioxide (0.30)
- residual sugar - total sulfur dioxide (0.40)
- residual sugar - density (0.84, denn gelöster Zucker beinflusst Dichte)
- residual sugar - alcohol (-0.45, denn Zucker gärt zu Alkohol)
- chlorides - alcohol (-0.36)
- free sulfur dioxide - total sulfur dioxide (0.62, denn free sulfur dioxide ist ein Teil der total sulfur dioxide)
- total sulfur dioxide - density(0.53)
- total sulfur dioxide - alcohol (-0,45)
- density - alcohol (-0,78)
- density - quality (-0.31)
- alcohol - quality (0.44)

### erwähnenswerte Zusammenhänge im Rotwein - Datensatz:
- fixed acidity - citric acid (0.67, denn citric acid ist ein Teil der fixed acidity)
- fixed acidity - density (0.67)
- fixed acidity - pH (-0.68, denn Menge an Säure im Wein bestimmt pH-Wert)
- volatile acidity - citric acid (-0.55)
- volatile acidity - quality (-0.39)
- citric acid - density (0.36)
- citric acid - pH (-0.54, , denn Menge an Säure im Wein bestimmt pH-Wert)
- residual sugar - density (0.36, denn gelöster Zucker beinflusst Dichte)
- chlorides - sulphates (0.37)
- free sulfur dioxide - total sulfur dioxide (0.67, denn free sulfur dioxide ist ein Teil der total sulfur dioxide)
- density - alcohol (-0.50)
- alcohol - quality (0.48)

Zusammenhänge von z. B. fixed acidity und citric acid (beide Datensätze) waren abzusehen, da citric acid ein Bestandteil der fixed acidity ist. Hingegen andere Korrelationen wie z. B. zwischen residual sugar und sulfur dioxide (Weißwein - Datensatz) sind neue Ergebnisse. Es gibt also sowohl klar absehbare als auch neue Korrelationen.

In [15]:
# Datensatz Rotwein
rw[kardinal_columns].corr()

fixed acidity  volatile acidity  citric acid  \
fixed acidity              1.000000         -0.256131     0.671703   
volatile acidity          -0.256131          1.000000    -0.552496   
citric acid                0.671703         -0.552496     1.000000   
residual sugar             0.114777          0.001918     0.143577   
chlorides                  0.093705          0.061298     0.203823   
free sulfur dioxide       -0.153794         -0.010504    -0.060978   
total sulfur dioxide      -0.113181          0.076470     0.035533   
density                    0.668047          0.022026     0.364947   
pH                        -0.682978          0.234937    -0.541904   
sulphates                  0.183006         -0.260987     0.312770   
alcohol                   -0.061668         -0.202288     0.109903   
quality                    0.124052         -0.390558     0.226373   

                      residual sugar  chlorides  free sulfur dioxide  \
fixed acidity               0.114777   0.093705            -0.153794   
volatile acidity            0.001918   0.061298            -0.010504   
citric acid                 0.143577   0.203823            -0.060978   
residual sugar              1.000000   0.055610             0.187049   
chlorides                   0.055610   1.000000             0.005562   
free sulfur dioxide         0.187049   0.005562             1.000000   
total sulfur dioxide        0.203028   0.047400             0.667666   
density                     0.355283   0.200632            -0.021946   
pH                         -0.085652  -0.265026             0.070377   
sulphates                   0.005527   0.371260             0.051658   
alcohol                     0.042075  -0.221141            -0.069408   
quality                     0.013732  -0.128907            -0.050656   

                      total sulfur dioxide   density        pH  sulphates  \
fixed acidity                    -0.113181  0.668047 -0.682978   0.183006   
volatile acidity                  0.076470  0.022026  0.234937  -0.260987   
citric acid                       0.035533  0.364947 -0.541904   0.312770   
residual sugar                    0.203028  0.355283 -0.085652   0.005527   
chlorides                         0.047400  0.200632 -0.265026   0.371260   
free sulfur dioxide               0.667666 -0.021946  0.070377   0.051658   
total sulfur dioxide              1.000000  0.071269 -0.066495   0.042947   
density                           0.071269  1.000000 -0.341699   0.148506   
pH                               -0.066495 -0.341699  1.000000  -0.196648   
sulphates                         0.042947  0.148506 -0.196648   1.000000   
alcohol                          -0.205654 -0.496180  0.205633   0.093595   
quality                          -0.185100 -0.174919 -0.057731   0.251397   

                       alcohol   quality  
fixed acidity        -0.061668  0.124052  
volatile acidity     -0.202288 -0.390558  
citric acid           0.109903  0.226373  
residual sugar        0.042075  0.013732  
chlorides            -0.221141 -0.128907  
free sulfur dioxide  -0.069408 -0.050656  
total sulfur dioxide -0.205654 -0.185100  
density              -0.496180 -0.174919  
pH                    0.205633 -0.057731  
sulphates             0.093595  0.251397  
alcohol               1.000000  0.476166  
quality               0.476166  1.000000

# Aufgabe 4.2

Erstellen Sie Visualisierungen Ihres Datensatzes - experimentieren Sie mit verschiedenen Diagrammtypen. Welche stellen sich als sinnvoll heraus, welche nicht und was fällt Ihnen auf? Können Sie neue Erkenntnisse gewinnen?

In [16]:
%matplotlib inline
import matplotlib
import matplotlib.pyplot as plt
import numpy as np
# numpy und pandas wurden oben schon importiert

quality_ww = ww["quality"]
alcohol_ww = ww["alcohol"]

ax.set_yticks(range(1, 11))
ax.set_xticks(range(16))

x = np.array(alcohol_ww)
y = np.array(quality_ww)

fig, ax = plt.subplots(figsize=(10,5))
ax.scatter(x,y)

fig.suptitle("Weinqualität in Abhängigkeit von Alkoholgehalt (Weißwein-Datensatz)")

plt.xlabel("Alkoholgehalt in %")
plt.ylabel("Weinqualität (0 - 10)")

NameError: name 'ax' is not defined

In [ ]:
%matplotlib inline
import matplotlib
import matplotlib.pyplot as plt
import numpy as np
# numpy und pandas wurden oben schon importiert

quality_ww = ww["quality"]
alcohol_ww = ww["alcohol"]

quality_rw = rw["quality"]
alcohol_rw = rw["alcohol"]

ax.set_yticks(range(1, 11))
ax.set_xticks(range(16))

x = np.array(alcohol_ww)
y = np.array(quality_ww)

x = np.array(alcohol_rw)
y = np.array(quality_rw)

fig, ax = plt.subplots(figsize=(10,5))
ax.scatter(x,y)

fig.suptitle("Weinqualität in Abhängigkeit von Alkoholgehalt (Rotwein-Datensatz)")

plt.xlabel("Alkoholgehalt in %")
plt.ylabel("Weinqualität (0 - 10)")

# Aufgabe 5.1
## Benford's Law

Analysieren Sie den von Ihnen gewählten Datensatz in Bezug auf Benfords Law - handelt es sich ggf. um synthetische Daten?

(???  Fragen: 1. Diagramme aus der Vorlesung irreführend, da Benford's Law nicht mit log [natürlicher Logarithmus], sondern mit log10 berechnet wird. 2. Soll bei 0.0... immer die erste Zahl ungleich null betrachtet werden?)

Vorgehen für Aufgabe:
1. Lass dir für jede Zelle in beiden Datensätzen die erste Zahl, die nicht null ist, ausgeben und in einen Dataframe/eine CSV-Datei eintragen.
2. Lass dir ausgeben, wie oft jede Zahl von 1 bis 9 in welcher Spalte des neuen Dataframes vorkommt. Rechne anhand dieser Werte die relative Häufigkeit der jeweiligen Zahl in der Spalte aus. Führe das für alle 12 Spalten durch. (Indexierung + Funktionen)
3. Vergleiche die erhaltenen relativen Häufigkeiten mit den theoretischen Häufigkeiten des Benford'schen Gesetzes (siehe Tabelle). Ist die Abweichung sehr groß, deutet das auf synthetische Daten hin.

In [29]:
benford = pd.DataFrame(range(1,10,1), columns=['N'])

benford["Benford - Wahrscheinlichkeit"] = np.log10(1+1/benford['N'])
display(benford)

N  Benford - Wahrscheinlichkeit
0  1                      0.301030
1  2                      0.176091
2  3                      0.124939
3  4                      0.096910
4  5                      0.079181
5  6                      0.066947
6  7                      0.057992
7  8                      0.051153
8  9                      0.045757

In [25]:
# Datensatz Weißwein

wwb = pd.read_csv('ww_benford.csv', sep=';')
wwb_final =  pd.read_csv('ww_benford_final.csv', sep=';')

def calculate_first_num(goal_file, from_dataframe, original_column):
    goal_file[f"1_{original_column}"] = ww[original_column].astype(str).str[0].astype(int)

col_list = ['fixed acidity', 'volatile acidity', 'citric acid', 'residual sugar',
            'chlorides', 'free sulfur dioxide', 'total sulfur dioxide', 'density',
            'pH', 'sulphates', 'alcohol', 'quality'
           ]

for column_name in col_list:
    calculate_first_num(wwb, ww, column_name)

display(wwb)

def calculate_benford(new_file, old_file, new_column):
    new_file[f"bf_{new_column}"] = np.log(old_file[column_name]+1)-np.log(old_file[column_name])

first_num_col_list = ['1_fixed acidity', '1_volatile acidity', '1_citric acid', '1_residual sugar',
            '1_chlorides', '1_free sulfur dioxide', '1_total sulfur dioxide', '1_density',
            '1_pH', '1_sulphates', '1_alcohol', '1_quality'
           ]

for column_name in first_num_col_list:
    calculate_benford(wwb_wwb, column_name)

display(wwb_final)

wwb_final.describe()

filler  1_fixed acidity  1_volatile acidity  1_citric acid  \
0       NaN                7                   0              0   
1       NaN                6                   0              0   
2       NaN                8                   0              0   
3       NaN                7                   0              0   
4       NaN                7                   0              0   
...     ...              ...                 ...            ...   
4893    NaN                6                   0              0   
4894    NaN                6                   0              0   
4895    NaN                6                   0              0   
4896    NaN                5                   0              0   
4897    NaN                6                   0              0   

      1_residual sugar  1_chlorides  1_free sulfur dioxide  \
0                    2            0                      4   
1                    1            0                      1   
2                    6            0                      3   
3                    8            0                      4   
4                    8            0                      4   
...                ...          ...                    ...   
4893                 1            0                      2   
4894                 8            0                      5   
4895                 1            0                      3   
4896                 1            0                      2   
4897                 0            0                      2   

      1_total sulfur dioxide  1_density  1_pH  1_sulphates  1_alcohol  \
0                          1          1     3            0          8   
1                          1          0     3            0          9   
2                          9          0     3            0          1   
3                          1          0     3            0          9   
4                          1          0     3            0          9   
...                      ...        ...   ...          ...        ...   
4893                       9          0     3            0          1   
4894                       1          0     3            0          9   
4895                       1          0     2            0          9   
4896                       1          0     3            0          1   
4897                       9          0     3            0          1   

      1_quality  
0             6  
1             6  
2             6  
3             6  
4             6  
...         ...  
4893          6  
4894          5  
4895          6  
4896          7  
4897          6  

[4898 rows x 13 columns]

C:\Users\z0055nwj\AppData\Local\Programs\Python\Python313\Lib\site-packages\pandas\core\arraylike.py:399: RuntimeWarning: divide by zero encountered in log
  result = getattr(ufunc, method)(*inputs, **kwargs)
C:\Users\z0055nwj\AppData\Local\Programs\Python\Python313\Lib\site-packages\pandas\core\arraylike.py:399: RuntimeWarning: divide by zero encountered in log
  result = getattr(ufunc, method)(*inputs, **kwargs)
C:\Users\z0055nwj\AppData\Local\Programs\Python\Python313\Lib\site-packages\pandas\core\arraylike.py:399: RuntimeWarning: divide by zero encountered in log
  result = getattr(ufunc, method)(*inputs, **kwargs)
C:\Users\z0055nwj\AppData\Local\Programs\Python\Python313\Lib\site-packages\pandas\core\arraylike.py:399: RuntimeWarning: divide by zero encountered in log
  result = getattr(ufunc, method)(*inputs, **kwargs)
C:\Users\z0055nwj\AppData\Local\Programs\Python\Python313\Lib\site-packages\pandas\core\arraylike.py:399: RuntimeWarning: divide by zero encountered in log
  result

filler  bf_1_fixed acidity  bf_1_volatile acidity  bf_1_citric acid  \
0       NaN            0.133531                    inf               inf   
1       NaN            0.154151                    inf               inf   
2       NaN            0.117783                    inf               inf   
3       NaN            0.133531                    inf               inf   
4       NaN            0.133531                    inf               inf   
...     ...                 ...                    ...               ...   
4893    NaN            0.154151                    inf               inf   
4894    NaN            0.154151                    inf               inf   
4895    NaN            0.154151                    inf               inf   
4896    NaN            0.182322                    inf               inf   
4897    NaN            0.154151                    inf               inf   

      bf_1_residual sugar  bf_1_chlorides  bf_1_free sulfur dioxide  \
0                0.405465             inf                  0.223144   
1                0.693147             inf                  0.693147   
2                0.154151             inf                  0.287682   
3                0.117783             inf                  0.223144   
4                0.117783             inf                  0.223144   
...                   ...             ...                       ...   
4893             0.693147             inf                  0.405465   
4894             0.117783             inf                  0.182322   
4895             0.693147             inf                  0.287682   
4896             0.693147             inf                  0.405465   
4897                  inf             inf                  0.405465   

      bf_1_total sulfur dioxide  bf_1_density   bf_1_pH  bf_1_sulphates  \
0                      0.693147      0.693147  0.287682             inf   
1                      0.693147           inf  0.287682             inf   
2                      0.105361           inf  0.287682             inf   
3                      0.693147           inf  0.287682             inf   
4                      0.693147           inf  0.287682             inf   
...                         ...           ...       ...             ...   
4893                   0.105361           inf  0.287682             inf   
4894                   0.693147           inf  0.287682             inf   
4895                   0.693147           inf  0.405465             inf   
4896                   0.693147           inf  0.287682             inf   
4897                   0.105361           inf  0.287682             inf   

      bf_1_alcohol  bf_1_quality  
0         0.117783      0.154151  
1         0.105361      0.154151  
2         0.693147      0.154151  
3         0.105361      0.154151  
4         0.105361      0.154151  
...            ...           ...  
4893      0.693147      0.154151  
4894      0.105361      0.182322  
4895      0.105361      0.154151  
4896      0.693147      0.133531  
4897      0.693147      0.154151  

[4898 rows x 13 columns]

C:\Users\z0055nwj\AppData\Local\Programs\Python\Python313\Lib\site-packages\pandas\core\nanops.py:1016: RuntimeWarning: invalid value encountered in subtract
  sqr = _ensure_numeric((avg - values) ** 2)
C:\Users\z0055nwj\AppData\Local\Programs\Python\Python313\Lib\site-packages\numpy\lib\_function_base_impl.py:4671: RuntimeWarning: invalid value encountered in subtract
  diff_b_a = subtract(b, a)
C:\Users\z0055nwj\AppData\Local\Programs\Python\Python313\Lib\site-packages\pandas\core\nanops.py:1016: RuntimeWarning: invalid value encountered in subtract
  sqr = _ensure_numeric((avg - values) ** 2)
C:\Users\z0055nwj\AppData\Local\Programs\Python\Python313\Lib\site-packages\numpy\lib\_function_base_impl.py:4671: RuntimeWarning: invalid value encountered in subtract
  diff_b_a = subtract(b, a)
C:\Users\z0055nwj\AppData\Local\Programs\Python\Python313\Lib\site-packages\pandas\core\nanops.py:1016: RuntimeWarning: invalid value encountered in subtract
  sqr = _ensure_numeric((avg - values) ** 

bf_1_fixed acidity  bf_1_volatile acidity  bf_1_citric acid  \
count         4898.000000            4898.000000       4898.000000   
mean             0.148667                    inf               inf   
std              0.031026                    NaN               NaN   
min              0.105361               0.693147          0.693147   
25%              0.133531                    NaN               NaN   
50%              0.154151                    NaN               NaN   
75%              0.154151                    NaN               NaN   
max              0.693147                    inf               inf   

       bf_1_residual sugar  bf_1_chlorides  bf_1_free sulfur dioxide  \
count          4898.000000          4898.0               4898.000000   
mean                   inf             inf                  0.333408   
std                    NaN             NaN                  0.168052   
min               0.105361             inf                  0.105361   
25%               0.182322             NaN                  0.223144   
50%               0.693147             NaN                  0.287682   
75%               0.693147             NaN                  0.405465   
max                    inf             inf                  0.693147   

       bf_1_total sulfur dioxide  bf_1_density      bf_1_pH  bf_1_sulphates  \
count                4898.000000   4898.000000  4898.000000     4898.000000   
mean                    0.567543           inf     0.298191             inf   
std                     0.220745           NaN     0.033579             NaN   
min                     0.105361      0.693147     0.287682        0.693147   
25%                     0.405465           NaN     0.287682             NaN   
50%                     0.693147           NaN     0.287682             NaN   
75%                     0.693147           NaN     0.287682             NaN   
max                     0.693147           inf     0.405465             inf   

       bf_1_alcohol  bf_1_quality  
count   4898.000000   4898.000000  
mean       0.463181      0.160318  
std        0.286078      0.023568  
min        0.105361      0.105361  
25%        0.105361      0.154151  
50%        0.693147      0.154151  
75%        0.693147      0.182322  
max        0.693147      0.287682

In [7]:
# Datensatz Rotwein

# --- Hilfsfunktion zur Extraktion der ersten nicht-null Ziffer ---
# Diese Funktion verwendet nur grundlegende String- und Schleifenoperationen.
def get_first_digit(number):
    """
    Extrahiert die erste nicht-null Ziffer einer Zahl (Integer oder Float).
    Berücksichtigt den Absolutwert der Zahl.
    Gibt np.nan zurück, wenn keine gültige erste Ziffer gefunden wird (z.B. für 0 oder NaN).
    """
    
    if number == 0:
        return np.nan

    # Den Absolutwert nehmen und in einen String umwandeln.
    s = str(abs(number))

    # Jedes Zeichen im String durchgehen
    for char in s:
        # Prüfen, ob das Zeichen eine Ziffer ist und nicht '0'
        if char.isdigit() and char != '0':
            return int(char) # Die erste gefundene nicht-null Ziffer als Integer zurückgeben
    
    # Falls keine nicht-null Ziffer gefunden wurde (z.B. bei "0.000")
    return np.nan

# --- Initialisierung des DataFrames für die Benford-Ergebnisse ---
# Wir erstellen einen neuen DataFrame, der die gleichen Index-Zeilen wie 'ww' hat.
benford_results_rw = pd.DataFrame(index=rw.index)

# --- Liste der Spalten, die verarbeitet werden sollen ---
columns_to_process = [
    'fixed acidity', 'volatile acidity', 'citric acid', 'residual sugar',
    'chlorides', 'free sulfur dioxide', 'total sulfur dioxide', 'density',
    'pH', 'sulphates', 'alcohol', 'quality'
]

# --- Hauptlogik: Iteration über Spalten und Zellen ---
# Wir gehen jede Spalte durch, die wir analysieren möchten
for col_name in columns_to_process:
    # Zuerst prüfen wir, ob die Spalte überhaupt in deinem DataFrame 'ww' existiert
    if col_name in rw.columns:
        # Eine temporäre Liste, um die Benford-Werte für die aktuelle Spalte zu sammeln
        benford_values_for_current_col = []

        # Jetzt gehen wir jeden einzelnen Wert (jede Zelle) in dieser Spalte durch
        for cell_value in rw[col_name]:
            # Die erste nicht-null Ziffer aus dem Zellenwert extrahieren
            first_digit = get_first_digit(cell_value)
            
            # Prüfen, ob eine gültige erste Ziffer gefunden wurde
            if pd.notna(first_digit):
                # Die Benford'sche Gesetz-Formel anwenden: P(d) = log10(1 + 1/d)
                # Dein ursprünglicher Code verwendete np.log, was der natürliche Logarithmus (Basis e) ist.
                # Benford's Gesetz wird traditionell mit dem Zehnerlogarithmus (Basis 10) formuliert.
                # Wenn du den Zehnerlogarithmus verwenden möchtest, ersetze np.log durch np.log10.
                # benford_value = np.log(first_digit + 1) - np.log(first_digit)
                # Oder für den Zehnerlogarithmus:
                benford_value = np.log10(first_digit + 1) - np.log10(first_digit)
                
                benford_values_for_current_col.append(benford_value)
            else:
                # Wenn keine gültige erste Ziffer gefunden wurde, fügen wir NaN hinzu
                benford_values_for_current_col.append(np.nan)

        # Nachdem wir alle Benford-Werte für die aktuelle Spalte gesammelt haben,
        # fügen wir sie als neue Spalte zum Ergebnis-DataFrame hinzu.
        # Leerzeichen im Spaltennamen werden durch Unterstriche ersetzt, damit es ein gültiger Name ist.
        new_col_name = f"Benford_{col_name.replace(' ', '_')}"
        benford_results_rw[new_col_name] = benford_values_for_current_col
    else:
        # Eine Warnung ausgeben, falls eine der gewünschten Spalten nicht im DataFrame 'rw' gefunden wurde
        print(f"Warnung: Spalte '{col_name}' wurde im DataFrame 'ww' nicht gefunden und übersprungen.")

# --- Anzeige des Ergebnis-DataFrames ---
print("\nDataFrame mit den Benford'schen Gesetz-Werten pro Zelle:")
print(benford_results_rw)
benford_results_rw.describe()


DataFrame mit den Benford'schen Gesetz-Werten pro Zelle:
      Benford_fixed_acidity  Benford_volatile_acidity  Benford_citric_acid  \
0                  0.057992                  0.057992                  NaN   
1                  0.057992                  0.051153                  NaN   
2                  0.057992                  0.057992             0.096910   
3                  0.301030                  0.176091             0.079181   
4                  0.057992                  0.057992                  NaN   
...                     ...                       ...                  ...   
1594               0.066947                  0.066947             0.051153   
1595               0.079181                  0.079181             0.301030   
1596               0.066947                  0.079181             0.301030   
1597               0.079181                  0.066947             0.301030   
1598               0.066947                  0.124939             0.096910   

     

Benford_fixed_acidity  Benford_volatile_acidity  Benford_citric_acid  \
count            1599.000000               1599.000000          1467.000000   
mean                0.099354                  0.097317             0.146347   
std                 0.092457                  0.046038             0.078087   
min                 0.045757                  0.045757             0.045757   
25%                 0.051153                  0.066947             0.096910   
50%                 0.057992                  0.079181             0.124939   
75%                 0.066947                  0.124939             0.176091   
max                 0.301030                  0.301030             0.301030   

       Benford_residual_sugar  Benford_chlorides  Benford_free_sulfur_dioxide  \
count             1599.000000        1599.000000                  1599.000000   
mean                 0.202083           0.090800                     0.182679   
std                  0.069793           0.080645                     0.100209   
min                  0.045757           0.045757                     0.045757   
25%                  0.176091           0.051153                     0.079181   
50%                  0.176091           0.057992                     0.176091   
75%                  0.301030           0.066947                     0.301030   
max                  0.301030           0.301030                     0.301030   

       Benford_total_sulfur_dioxide  Benford_density   Benford_pH  \
count                   1599.000000      1599.000000  1599.000000   
mean                       0.159647         0.058689     0.125831   
std                        0.093513         0.055998     0.006904   
min                        0.045757         0.045757     0.096910   
25%                        0.079181         0.045757     0.124939   
50%                        0.124939         0.045757     0.124939   
75%                        0.301030         0.045757     0.124939   
max                        0.301030         0.301030     0.176091   

       Benford_sulphates  Benford_alcohol  Benford_quality  
count        1599.000000      1599.000000      1599.000000  
mean            0.078897         0.192495         0.072221  
std             0.045251         0.126215         0.009979  
min             0.045757         0.045757         0.051153  
25%             0.057992         0.045757         0.066947  
50%             0.066947         0.301030         0.066947  
75%             0.079181         0.301030         0.079181  
max             0.301030         0.301030         0.124939

In beiden Datensätzen ist 